2024/incoming/pixie-the-elf-picks-an-artist-exploring-perl-s-new-class-syntax.pod

```diff
@@ -0,0 +1,267 @@
+
+=end code
+
+Looking back over the code, Pixie added a TODO comment reminding herself
+that she should really deal with the case when the environment variables
+weren't set. She'd do that... one day.
+
+Next, Pixie came to the methods for her class. There were three main
+differences to the OO Perl she had written before. Firstly, she could define
+methods with a C<method> keyword. Secondly, each method has access to the
+class fields - and you can just treat them like variables. And, finally, each
+method automatically has access to a C<$self> variable - without having to
+read it from the arguments!
+
+The first method was simple:
+
+=begin code
+
+    method run {
+      $self->getartists;
+      $self->pickartist;
+      $self->render;
+    }
+
+=end code
+
+Pixie was following good programming design practice here. She broke the
+problem down into simpler steps. One to get the artists from the LastFM
+API, one to pick an artist to recommend and one to display the results
+to the user.
+
+The next method was the most complex:
+
+=begin code
+
+    method getartists {
+      my $data;
+      my $page = 1;
+      do {
+        $data = $lastfm->request_signed(
+          method => $method,
+          user   => $username,
+          page   => $page,
+        );
+        for (@{$data->{topartists}{artist}}) {
+          next if $_->{playcount} > $max;
+          next if $_->{playcount} < $min;
+          push @artists, $_;
+        }
+        ++$page;
+      } until $data->{topartists}{artist}[-1]{playcount} < $min;
+
+      die "No artists with between $min and $max plays for $username\n"
+        unless @artists;
+    }
+
+=end code
+
+This is where Pixie makes the actual call to the LastFM API. Most of the
+complexity is caused by LastFM paging the data it gives you, so you might
+need to make multiple calls in order to get all the data you want.
+
+But at the end of this method, the C<@artists> field contains one element
+for each artist the user has listened to between the minumum and maximum
+requested number of times.
+
+The last two methods are so simple, that Pixie hesitated to even make them
+methods:
+
+=begin code
+
+    method pickartist {
+      $artist = $artists[ rand @artists ];
+    }
+
+    method render {
+      say "$artist->{name} ($artist->{playcount})";
+    }
+
+=end code
+
+Pixie's code is available on GitHub -
+L<github.com/davorg/pickanartist|https://github.com/davorg/pickanartist>
+
+=head2 Embracing Modern Perl
+
+Pixie is happy with the way the code worked out. She thinks that the new
+syntax makes her code far easier to read, maintain and extend. She
+particularly like the way that the fields looked and acted like ordinary
+variables. This made a lot of code far simpler. Under "classic" Perl class
+syntax (and even with Moo or Moose) you would print an attribute with code
+like this:
+
+    say "The name is ", $self->name;
+
+But the new syntax simplifies that to:
+
+    say "The name is $name";
+
```

Comment:
Upon closer reading, I'm not actually sure I understand this part. 🤔

Reply:
With "classic" Perl OO (or even with Moo/se) when you want to access the value of an attribute inside a method you can either use the accessor method (`$self->name`) or directly access the blessed hash (`$self->{name}`). With the accessor method, you're now dealing with a subroutine call, so you can't easily interpolate it into a string. With the direct hash access, you're breaking encapsulation, which is bad practice and will burn you if the internal representation were to change.

With "perlclass", inside a method, an attribute (a field) looks and acts just like a variable. Interpolation Just Works and you don't need to care about the internal representation.

I can pad out the description a bit if you think it's necessary.

Reply:
If you could, that would be really helpful, I think. I searched for `$name` in the article and couldn't find any other mention of it, so it wasn't immediately clear what was going on in the context of the preceding code.

Reply:
If you could branch off #487 (publish/2024-12-08), that would make for a better chance to avoid a merge conflict.

Reply:
Also, if you could think of an image to include (like a screen shot of recommendations in action) that would make for a catchier social media preview, but that's not necessary by any means.

Reply:
Oh, yeah. It would probably have helped if I had used one of the fields that actually exists in the class!

Reply:
I mean, command line programs are famously unphotogenic :-)

But have a picture of Pixie listening to music (I quite understand if you don't want to use AI-generated images).

Reply:
https://github.com/perladvent/Perl-Advent/pull/488/